src/credentials/tests/TestDeviceAttestationConstruction.cpp

```diff
@@ -0,0 +1,159 @@
+    ByteSpan vendorReservedArray[] = { ByteSpan(vendorReserved1), ByteSpan(vendorReserved3) };
+    uint16_t vendorId              = 0xbeef;
+    uint16_t profileNum            = 0xdead;
+
+    attestationElementsLen = sizeof(certificationDeclaration) + sizeof(attestationNonce) + sizeof(timestamp) +
+        sizeof(vendorReserved1) + sizeof(vendorReserved3) + sizeof(uint64_t) * 5;
+    attestationElements.Alloc(attestationElementsLen);
+    NL_TEST_ASSERT(inSuite, attestationElements);
+
+    {
+        MutableByteSpan attestationElementsSpan(attestationElements.Get(), attestationElementsLen);
+
+        // test wrong size nonce
+        err = ConstructAttestationElements(
+            ByteSpan(certificationDeclaration), ByteSpan(attestationNonce, sizeof(attestationNonce) - 1), timestamp, ByteSpan(),
+            vendorReservedArray, ArraySize(vendorReservedArray), vendorId, profileNum, attestationElementsSpan);
+        NL_TEST_ASSERT(inSuite, err == CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+        // test with missing mandatory TLV entries
+        err = ConstructAttestationElements(ByteSpan(), ByteSpan(attestationNonce), timestamp, ByteSpan(), vendorReservedArray,
+                                           ArraySize(vendorReservedArray), vendorId, profileNum, attestationElementsSpan);
+        NL_TEST_ASSERT(inSuite, err == CHIP_ERROR_INVALID_ARGUMENT);
+
+        // test with missing mandatory TLV entries
+        err =
+            ConstructAttestationElements(ByteSpan(certificationDeclaration), ByteSpan(), timestamp, ByteSpan(), vendorReservedArray,
+                                         ArraySize(vendorReservedArray), vendorId, profileNum, attestationElementsSpan);
+        NL_TEST_ASSERT(inSuite, err == CHIP_ERROR_INVALID_ARGUMENT);
+
+        err = ConstructAttestationElements(ByteSpan(certificationDeclaration), ByteSpan(attestationNonce), timestamp, ByteSpan(),
+                                           vendorReservedArray, ArraySize(vendorReservedArray), vendorId, profileNum,
+                                           attestationElementsSpan);
+        NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+        attestationElementsLen = attestationElementsSpan.size();
+    }
+
+    ByteSpan certificationDeclarationSpan;
+    ByteSpan attestationNonceSpan;
+    uint32_t timestampDeconstructed;
+    ByteSpan firmwareInfoSpan;
+    ByteSpan vendorReservedDeconstructed[2];
+    uint16_t vendorIdDeconstructed;
+    uint16_t profileNumDeconstructed;
+
+    err =
+        DeconstructAttestationElements(ByteSpan(attestationElements.Get(), attestationElementsLen), certificationDeclarationSpan,
+                                       attestationNonceSpan, timestampDeconstructed, firmwareInfoSpan, vendorReservedDeconstructed,
+                                       ArraySize(vendorReservedDeconstructed), vendorIdDeconstructed, profileNumDeconstructed);
```

Comment:
This is a round-trip test. It does not externally validate that construction is done properly. 

Need to add tests to cover for parser doing:

1- At least one TLV tag out of order during parsing
2- firmware info present/missing
3- Being able to generate the exact byte stream from the spec test vector, and parsing the same byte string from test vector, without going through round-tripping.

Reply:
will update